tools/rpkg/src/statement.cpp

```diff
@@ -449,14 +449,42 @@ struct AppendableRList {
 	RProtector r;
 };
 
-// Turn a DuckDB materialized result set into an Arrow Table
-SEXP duckdb_execute_arrow(QueryResult *result) {
+struct RQueryResult {
+	unique_ptr<QueryResult> result;
+};
+
+bool FetchArrowChunk(QueryResult *result, AppendableRList &batches_list, ArrowArray &arrow_data,
+                     ArrowSchema &arrow_schema, SEXP &batch_import_from_c, SEXP &arrow_namespace) {
+	if (result->type == QueryResultType::STREAM_RESULT) {
+		auto stream_result = (StreamQueryResult *)result;
+		if (!stream_result->is_open) {
+			return false;
+		}
+	}
+	unique_ptr<DataChunk> data_chunk = result->Fetch();
+	if (!data_chunk || data_chunk->size() == 0) {
+		return false;
+	}
+	result->ToArrowSchema(&arrow_schema);
+	data_chunk->ToArrowArray(&arrow_data);
+	batches_list.PrepAppend();
+	batches_list.Append(RApi::REvalRerror(batch_import_from_c, arrow_namespace));
+	return true;
+}
+
+// Turn a DuckDB result set into an Arrow Table
+SEXP RApi::DuckDBExecuteArrow(SEXP query_resultsexp, SEXP streamsexp, SEXP vector_per_chunksexp,
+                              SEXP return_tablesexp) {
 	RProtector r;
+	RQueryResult *query_result_holder = (RQueryResult *)R_ExternalPtrAddr(query_resultsexp);
+	auto result = query_result_holder->result.get();
 	// somewhat dark magic below
 	SEXP arrow_name_sexp = r.Protect(RApi::StringsToSexp({"arrow"}));
 	SEXP arrow_namespace_call = r.Protect(Rf_lang2(Rf_install("getNamespace"), arrow_name_sexp));
 	SEXP arrow_namespace = r.Protect(RApi::REvalRerror(arrow_namespace_call, R_GlobalEnv));
-
+	bool stream = LOGICAL_POINTER(streamsexp)[0] != 0;
```

Comment:
perhaps add some checks here, `streamsexp` and `vector_per_chunksexp` need to be a LOGICAL of length 1, `vector_per_chunksexp` needs to be numeric of length 1. 
e.g.
````C++
if (TYPEOF(streamsexp) != LGLSXP || LENGTH(streamsexp) != 1) {
  Rf_error("stream parameter needs to be single-value logical");
}
````